bootstrap/tracing.go

```diff
@@ -0,0 +1,167 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/buildkite/agent/v3/env"
+	"github.com/buildkite/agent/v3/tracetools"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/exp/slices"
+	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+func noopStopper() {}
+
+func (b *Bootstrap) startTracing(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	switch b.Config.TracingBackend {
+	case tracetools.BackendDatadog:
+		// Newer versions of the tracing libs print out diagnostic info which spams the
+		// Buildkite agent logs. Disable it by default unless it's been explicitly set.
+		if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
+			os.Setenv("DD_TRACE_STARTUP_LOGS", "false")
+		}
+
+		return b.startTracingDatadog(ctx)
+
+	case tracetools.BackendNone:
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+
+	default:
+		b.shell.Commentf("An invalid tracing backend was provided: %q. Tracing will not occur.", b.Config.TracingBackend)
+		b.Config.TracingBackend = tracetools.BackendNone // Ensure that we don't do any tracing after this, some of the stuff in tracetools uses the bootstrap's tracking backend
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+	}
+}
+
+func (b *Bootstrap) tracingResourceName() string {
+	label, ok := b.shell.Env.Get("BUILDKITE_LABEL")
+	if !ok {
+		label = "job"
```

Comment:
Should this branch also include the organization and pipeline slugs? I'm pretty sure it used to.

Reply:
oh it did! lost in rebase hell, thanks for pointing this out

Reply:
oh, wait, these are just a couple of lines down now, [here](https://github.com/buildkite/agent/pull/1631/files/fec0ca97087532b9984909a969f50d7e48d9183c#diff-f80935c46a08d736109e6791c2b0c9a1b004a9c196a8a149b3a70d5c2d3d40c0R53)

Reply:
Oh, I just can't read. In my head this returned early. Never mind. 🙈